### ISSUE TITLE
DOC: basis functions and weak formulation for those without a background in finite elements

### ISSUE THREAD

**Issue by gdmcbain** [kinnala/scikit-fem#423]
> For someone who does not have a background in finite elements, they may not understand what basis functions or weak formulation means. (#422)

**Comment by gdmcbain** [comment]
My initial thoughts are that (at present), scikit-fem is either for:
* those who do have a background in finite elements and know what basis functions and weak formulation mean, or
* for internal use in other packages in which the end-user need not have background in finite elements but doesn't have to look inside the functions in that package.

In the first case, the design idea of scikit-fem is that it adheres closely to the standard fundamental notions of the finite element method as understood and accepted by those with a background in finite elements, e.g. as set forth in the books of Braess, Ern & Guermond, Elman, Silvester, & Wathen, &c., &c., or as in other codes like FreeFEM, libMesh, FEniCS, &c.,, &c.; and expresses them in Python (in modern Python, Pythonically, and 'thinly' in the SciPy software stack).  scikit-fem is deliberately not (at present) innovative at the level of the basic finite element notions, anyone having read one of those books or used one of those codes shouldn't have much trouble seeing what MeshTet, ElementTetP1, InteriorBasis, BilinearForm, LinearForm, &c., are intending.

In the second case, the writers of those parts of the downstream packages need to be in the first case, but the users of those packages might not even know that scikit-fem or finite elements are being used at all.  This is like someone might use [`scipy.integrate.solve_ivp`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html) without understanding the theory & nomenclature of Runge&ndash;Kutta methods (if indeed that's what they use: that I don't actually know, which is just the point).  Actually more I'm thinking of higher-level application-specific packages like the battery-simulator, but I'm guessing that it would be possible to write a drop-in re-implementation of `scipy.integrate.solve_bvp` using scikit-fem without the user having to know about basis functions or weak formulations.

These two use-cases are both important but have to be distinguished from each other.  I think it is possible for a single package to serve both cases without compromise, largely thanks to the design of the Python language and the SciPy stack.

I'd say that (at present) scikit-fem is not for the in-between case of the end-user who wants to set up and solve a general mechanical stress problem or heat transfer problem using finite element analysis but without knowledge of finite elements.

Anyway, these are just my initial thoughts, without rechecking what was written in the submission.  Were they covered there?  Or do they need to be emphasized or added?  I'll recheck the text later.

**Comment by kinnala** [comment]
Some approaches:
- provide references for those who do not know the terminology but are interested to learn more
- try to explain or clarify the meaning, e.g., via an example

**Comment by gdmcbain** [comment]
Good ideas.

For the first, I looked in the books of Braess and Ern & Guermond; I wouldn't recommend their explanations of basis functions or weak formulation to those who hadn't heard the terms before.  Those two are always on my desk; I'll check the shelf for others.

Then I looked in https://kinnala.github.io/scikit-fem-docs/.  The first line contains a link to https://en.wikipedia.org/wiki/Finite_element_method, which is good (though a horrible nomenclature on that page is that they use P1 and P2 to denote two boundary value problems&mdash;of course I initially confused them with the piecewise-polynomial Lagrangian elements of degree 1 and 2, denoted $\mathcal P_1$ and $\mathcal P_2$ by Braess).  Both concepts are reasonably well explained there, I think.

Thinking back to where I learned these concepts myself, it was probably in the tutorials to other finite element packages.  There's a lot of redundancy to these.  (This repetition was part of the motivation for starting gdmcbain/fenics-tuto-in-skfem; one can pretty much read the FEniCS tutorial which is nine-tenths about the finite element method and then drop in the one-tenth about the implementation and examples in skfem.)  There used to be a walkthrough at kinnala/scikit-fem-docs which was good, from memory, but I see that it's been retracted, so we can't refer to that.
